# Vaihe 1

### Kiinnostavat tutkimuskysymykset
1. Mitkä genret ovat olleet suosittuja milloin?
2. Mikä näyttelijöiden yhdistelmä tuottaa suosittuja elokuvia?
3. Korrelaatio arvioiden ja tuoton välillä. Voiko elokuva olla huonosti arvioitu mutta silti tuottoisa tai toisin päin?
4. Jos tekisimme elokuvan, millainen sen kannattaa olla, että se on tuottoisa ja suosittu?

# Vaihe 2

### Aineiston kuvaus
Aineisto löydettiin Kagglesta ja se sisältää tietoa elokuvista IMDB palvelussa.

Linkki aineistoon: [LINKKI](https://www.kaggle.com/datasets/ashpalsingh1525/imdb-movies-dataset)


In [7]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
# set display options
pd.set_option('display.max_colwidth', None)

dataset = pd.read_csv('Datasets/imdb_movies.csv')

# TODO: Fixaa tää silleen et selitetään otsikoiden tarkoitus
# Print dataframe columns
print("Aineiston otsikot:")
print(dataset.columns)
print("Otsikoiden tyypit:")
print(dataset.dtypes)
print("Aineiston muoto:")
print(dataset.shape)
amount_of_rows_pre = dataset.shape[0]


Aineiston otsikot:
Index(['names', 'date_x', 'score', 'genre', 'overview', 'crew', 'orig_title',
       'status', 'orig_lang', 'budget_x', 'revenue', 'country'],
      dtype='object')
Otsikoiden tyypit:
names          object
date_x         object
score         float64
genre          object
overview       object
crew           object
orig_title     object
status         object
orig_lang      object
budget_x      float64
revenue       float64
country        object
dtype: object
Aineiston muoto:
(10178, 12)


#### Aineiston otsikot:
- names:        Elokuvan nimi englanniksi
- date_x:       Elokuvan julkaisupäivämäärä
- score:        IMDB-käyttäjien antaman pistemäärän keskiarvo 
- genre:        Elokuvan genre
- overview:     Elokuvan kuvaus englanniksi
- crew:         Elokuvan näyttelijät ja heidän esittämät hahmot. esim. (näyttelijä, hahmo, näyttelijä, hahmo)
- orig_title:   Elokuvan nimi alkuperäisellä kielellä
- status:       Elokuvan tilanne. (Kaikki havainnot ovat julkaistu)
- orig_lang:    Elokuvan alkuperäiskieli
- budget_x:     Elokuvan budjetti (USD)
- revenue:      Elokuvan bruttotuotto (USD)
- country:      Elokuvan julkaisumaa

Aineistossa on siis 12 otsikkoa ja 10178 havaintoa.


Osa aineiston datasta on epäilyttävää. Esimerkiksi 144 elokuvista on saanut 0.0 arvostelun, ja niiden kaikkien tuotto on tasan $175,269,998.80

Aineiston julkaisija on ilmeisesti käyttänyt jotain knn keinoa aineiston täyttämiseen. Tämä täytyy ottaa huomioon myöhemmin.



# Vaihe 3
## Aineiston esikäsittely

#### Huonojen havaintojen poisto

In [8]:
# Tässä poistetaan ne havainnot joissa on tyhjiä arvoja tai arvostelu on 0.0
dataset = dataset.dropna()
dataset = dataset[dataset['score'] != 0.0]
print(dataset.shape)
amount_of_rows_post = dataset.shape[0]
print("Poistettu " + str(amount_of_rows_pre - amount_of_rows_post) + " riviä, jossa oli tyhjiä arvoja tai arvostelu oli 0.0")

(9889, 12)
Poistettu 289 riviä, jossa oli tyhjiä arvoja tai arvostelu oli 0.0


#### Avataan näyttelijät
Emme ole kiinnostuneita näyttelijöiden hahmoista, joten poistetaan ne aineistosta.

In [9]:
print("1. elokuvan näyttelijät:")
print(dataset['crew'].head(1))

1. elokuvan näyttelijät:
0    Michael B. Jordan, Adonis Creed, Tessa Thompson, Bianca Taylor, Jonathan Majors, Damien Anderson, Wood Harris, Tony 'Little Duke' Evers, Phylicia Rashād, Mary Anne Creed, Mila Davis-Kent, Amara Creed, Florian Munteanu, Viktor Drago, José Benavidez Jr., Felix Chavez, Selenis Leyva, Laura Chavez
Name: crew, dtype: object


In [10]:
# Tässä avataan näyttelijät. Näyttelijän hahmo on listattu näyttelijän jälkeen. (esim. Tom Cruise, Ethan Hunt, ...)
dataset['crew'] = dataset['crew'].str.split(', ')
# Poista joka toinen alkio
dataset['crew'] = dataset['crew'].apply(lambda x: x[::2])
print("1. elokuvan näyttelijät:")
print(dataset['crew'].head(1))

1. elokuvan näyttelijät:
0    [Michael B. Jordan, Tessa Thompson, Jonathan Majors, Wood Harris, Phylicia Rashād, Mila Davis-Kent, Florian Munteanu, José Benavidez Jr., Selenis Leyva]
Name: crew, dtype: object


Crew -sarake sisältää siis nyt vain näyttelijät ilman hahmoja

# Vaihe 4
### Aineiston analysointi

In [37]:
#  JOKAISESSA ANALYSOINTIKOHDASSA TEHDÄÄN AINEISTOSTA KOPIO, ETTEI MYÖHEMMÄT KOHDAT MUUTU

#### Mitkä genret ovat suosittuja milloin

In [11]:
dataset_genre = dataset.copy()

# Extract year from the release date
dataset_genre['date_x'] = pd.to_datetime(dataset_genre['date_x'])
dataset_genre['year'] = dataset_genre['date_x'].dt.year

# Split genres
dataset_genre['genre'] = dataset_genre['genre'].str.split(', ')
dataset_genre = dataset_genre.explode('genre')

# Group by genre and year
grouped = dataset_genre.groupby(['genre', dataset_genre['year'].dt.year]).size().unstack()

# Plot the results
grouped.plot(kind='bar', stacked=True, figsize=(15, 10))
plt.title('Poplarity of movie genres over the years')
plt.ylabel('Amount of movies')
plt.xlabel('Year')
plt.legend(title='Genre', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.tight_layout()
plt.show()

AttributeError: Can only use .dt accessor with datetimelike values

#### Mikä näyttelijöiden yhdistelmä tuottaa suosittuja elokuvia?

#### Korrelaatio arvioiden ja tuoton välillä.

In [63]:
import pandas as pd
import seaborn as sns

# Kopioi aineisto
dataset_correlation = dataset.copy()

# Laske profit ja ROI
dataset_correlation['profit'] = dataset_correlation['revenue'] - dataset_correlation['budget_x']
dataset_correlation['roi'] = dataset_correlation['profit'] / dataset_correlation['budget_x']
# poista ylimääräiset välilyönnit
dataset_correlation['genre'] = dataset_correlation['genre'].str.strip().str.split(r'\s*,\s*')
# Erotellaan genret erillisiksi riveiksi
df = dataset_correlation.explode('genre')
genre_stats = df.groupby('genre')[['score', 'profit']].mean().sort_values(by='profit', ascending=True)
# Tulostetaan tulos
print(genre_stats)




                     score        profit
genre                                   
War              69.845324  1.328108e+08
Western          68.530769  1.382839e+08
Crime            65.802067  1.394526e+08
History          69.820574  1.427855e+08
Thriller         63.207591  1.483761e+08
Horror           60.352406  1.538480e+08
Mystery          64.942556  1.689148e+08
Drama            67.057395  1.698814e+08
Action           64.469049  1.835957e+08
Comedy           64.643963  1.863774e+08
Romance          64.402742  1.942138e+08
Science Fiction  63.984763  2.049001e+08
Adventure        65.859433  2.173471e+08
Music            70.014760  2.198567e+08
Fantasy          66.384446  2.265352e+08
Family           66.611908  2.384640e+08
Animation        69.522853  2.891327e+08
TV Movie         65.757282  3.834781e+08
Documentary      68.930851  4.484902e+08


### Voiko elokuva olla huonosti arvioitu mutta silti tuottoisa tai toisin päin?


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Laske keskiarvot
mean_score = dataset['score'].mean()
mean_revenue = dataset['revenue'].mean()

# Luo kategoriat
dataset['category'] = 'Muu'
dataset.loc[(dataset['score'] < mean_score) & (dataset['revenue'] > mean_revenue), 'category'] = 'Huonosti arvioitu, mutta tuottoisa'
dataset.loc[(dataset['score'] > mean_score) & (dataset['revenue'] < mean_revenue), 'category'] = 'Hyvin arvioitu, mutta vähän tuottoa'

# Poista 'Muu' kategoria
filtered_dataset = dataset[dataset['category'] != 'Muu']
# Tulosta kategorioiden jakauma
print(filtered_dataset['category'].value_counts())

# Visualisoi kategoriat
plt.figure(figsize=(10, 6))
sns.scatterplot(x='score', y='revenue', hue='category', data=filtered_dataset, palette='Set1', alpha=0.7)
plt.axhline(mean_revenue, color='red', linestyle='--', label='Keskiarvo (tuotto)')
plt.axvline(mean_score, color='blue', linestyle='--', label='Keskiarvo (arvosana)')
plt.title('Elokuvien luokittelu arvioiden ja tuoton mukaan')
plt.xlabel('IMDB-arvosana')
plt.ylabel('Tuotto (USD)')
plt.legend()
plt.show()

In [64]:
# Huonosti arvioitu, mutta tuottoisin elokuva
max_revenue_low_score = dataset[dataset['category'] == 'Huonosti arvioitu, mutta tuottoisa'].nlargest(1, 'revenue')

print("Huonosti arvioitu, mutta tuottoisin elokuva:")
print(f"Arvosana: {max_revenue_low_score['score'].values[0]}")
print(f"Tuotto: {max_revenue_low_score['revenue'].values[0]} USD")
print(f"Nimi: {max_revenue_low_score['names'].values[0]}")
print("\n")

# Etsi "Hyvin arvioitu, mutta vähiten tuottoinen elokuva"
min_revenue_high_score = dataset[dataset['category'] == 'Hyvin arvioitu, mutta vähän tuottoa'].nsmallest(1, 'revenue')


print("Hyvin arvioitu, mutta vähiten tuottoinen elokuva:")
print(f"Arvosana: {min_revenue_high_score['score'].values[0]}")
print(f"Tuotto: {min_revenue_high_score['revenue'].values[0]} USD")
print(f"Nimi: {min_revenue_high_score['names'].values[0]}")

Huonosti arvioitu, mutta tuottoisin elokuva:
Arvosana: 44.0
Tuotto: 1515100211.0 USD
Nimi: The Avengers


Hyvin arvioitu, mutta vähiten tuottoinen elokuva:
Arvosana: 78.0
Tuotto: 0.0 USD
Nimi: Prey


#### Mitä visualisaatiosta voi lukea?
Huonosti arvioitu, mutta tuottoisa:
Nämä elokuvat saavat huonoja arvosanoja, mutta ne tuottavat silti paljon rahaa. Tämä voi viitata siihen, että elokuva on kaupallisesti menestys, mutta ihmiset eivät pitäneet siitä.

Hyvin arvioitu, mutta vähän tuottoa:

Nämä elokuvat saavat hyviä arvosanoja, mutta ne eivät tuota paljon rahaa. Tämä voi tarkoittaa, että elokuva on arvostettu ihmiseten keskuudessa, mutta se ei ole saavuttanut laajaa yleisöä.

#### Jos tekisimme elokuvan, millainen sen kannattaa olla, että se on tuottoisa ja suosittu?

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Copy the dataset
dataset_profitable = dataset.copy()

# Drop rows with missing values
dataset_profitable = dataset_profitable.dropna()
dataset_profitable = dataset_profitable[dataset_profitable['score'] != 0.0]

# Extract year from the release date
dataset_profitable['date_x'] = pd.to_datetime(dataset_profitable['date_x'])
dataset_profitable['year'] = dataset_profitable['date_x'].dt.year

# Convert categorical variables to dummy variables
dataset_profitable = pd.get_dummies(dataset_profitable, columns=['genre', 'country', 'orig_lang'], drop_first=True)

# Define the features and target variable
X = dataset_profitable.drop(columns=['names', 'date_x', 'overview', 'crew', 'orig_title', 'status', 'revenue'])
y = dataset_profitable['revenue'] > dataset_profitable['revenue'].mean()

# Split the dataset into a training set and a testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Predict
y_pred = model.predict(X_test)

# Evaluate
print("Logistic Regression Metrics:")
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Classification Report:')
print(classification_report(y_test, y_pred))
print('Confusion Matrix:')
print(confusion_matrix(y_test, y_pred))



# Vaihe 5
### Analyysin tulkinta

# Vaihe 6
### Tulosten hyödyntäminen